feature(views): Nearly all plugin static resources are in views

mrclay opened the pull request:
Makes elgg.get_simplecache_url() match the PHP version.
Associates .swf files with the right Content-Type.
groups/icon.php sends the right Content-Type for gifs.

(this is a follow up to #8513)

mrclay commented:
Should we craft a convention for view names of images?

mrclay commented:
Alternately we could leave the files in place for BC and just use `views.php` files to do this.

ewinslow commented on mod/ckeditor/views.php:
```diff
@@ -0,0 +1,7 @@
+return [
+	'default' => [
+		'js/ckeditor.js' => __DIR__ . '/vendors/ckeditor/ckeditor.js',
+		'js/jquery.ckeditor.js' => __DIR__ . '/vendors/ckeditor/adapters/jquery.js',
```
Love this.

ewinslow commented on mod/ckeditor/views/default/js/elgg/ckeditor.js:
```diff
@@ -3,7 +3,7 @@ define(function(require) {
 	var CKEDITOR = require('ckeditor');
 
-	CKEDITOR.plugins.addExternal('blockimagepaste', elgg.get_site_url() + 'mod/ckeditor/views/default/js/elgg/ckeditor/blockimagepaste.js', '');
+	CKEDITOR.plugins.addExternal('blockimagepaste', elgg.get_simplecache_url('js/elgg/ckeditor/blockimagepaste.js'), '');
```
I've been using require.toUrl for this. Haven't decided which one is "better"

ewinslow commented on mod/groups/start.php:
```diff
@@ -329,7 +329,7 @@ function groups_set_icon_url($hook, $type, $url, $params) {
 		return "groupicon/$group->guid/$size/$icontime.jpg";
 	}
 
-	return "mod/groups/graphics/default{$size}.gif";
```
I suspect the number of plugins pointing to `mod/plugin/someimage.jpg` is pretty high... No way that I know to forward with redirects. Really like pushing everything into views but this might be a case for views.php -- all gain no pain as far as I can tell...

ewinslow replied:
One problem is that if someone tries to override the image in a view, but the direct references are left, that could actually feel broken because it will show the "overridden" image.

mrclay replied:
^ Can you explain this more?

ewinslow replied:
Suppose a plugin points to

/mod/groups/graphics/defaultlarge.gif

In 2.0, a theme tries to override this new view we've created. The old image will still show in some cases because the plugin is pointing directly to the filesystem location.

mrclay replied:
Yep, something for the upgrading docs

ewinslow commented on mod/zaudio/views/default/js/elgg/zaudio.js:
```diff
@@ -7,7 +7,7 @@ define(function (require) {
 	var i = 0;
 
-	AudioPlayer.setup(elgg.get_site_url() + "mod/zaudio/audioplayer/player.20150521.swf", {width: 290});
+	AudioPlayer.setup(elgg.get_simplecache_url('AudioPlayer.swf'), {width: 290});
```
If you use require.toUrl, you can get rid of the `elgg` dependency entirely from this file, which seems like a Good Thing.

mrclay replied:
Won't this request `js/AudioPlayer.swf.js`?

ewinslow replied:
Based on this snippet from the docs, I don't think so:

```js
define(["require"], function(require) {
    var cssUrl = require.toUrl("./style.css");
});
```

mrclay replied:
I had to use `require.toUrl('../../AudioPlayer.swf')`. That seems more confusing to me.

ewinslow replied:
Ah, well if it only does relative URLs then yes, that seems less than ideal...

mrclay committed:
feature(views): Nearly all plugin static resources are in views

Makes elgg.get_simplecache_url() match the PHP version.
Associates .swf files with the right Content-Type.
groups/icon.php sends the right Content-Type for gifs.

mrclay commented:
Redone without moving image files.

ewinslow commented:
This looks good to me. The more examples of this new feature the better :)

mrclay commented on mod/file/views.php:
```diff
@@ -0,0 +1,16 @@
+	}
+
+	return $ret;
+});
```
@ewinslow should we do something like this for core `_graphics`?

ewinslow replied:
I do still prefer core the way it is if we can keep it that way...

Actually I just had a thought, what if we used .htaccess in plugins to forward through simplecache? Wouldn't work for nginx but could be a fix for the theming problem I described above for apache, which seems like the most common configuration.

Might be a bad idea to have such different behavior depending on web server though...

mrclay replied:
Wouldn't be reliable either, due to RewriteBase

mrclay replied:
I made #8550 to remind me to replace this

ewinslow commented:
@mrclay Anything left here?

mrclay commented:
Nope.

ewinslow commented:
Sweet! Thanks for doing that
